File: day14/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	mem := map[int]int{}
	mem2 := map[int]int{}

	maskZeros := 0
	maskOnes := 0
	maskFloats := 0
	maskPos := []int{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		s := strings.Split(scanner.Text(), " = ")
		if len(s) != 2 {
			log.Fatal("Invalid line")
		}
		if s[0] == "mask" {
			maskZeros, maskOnes, maskFloats, maskPos = processMask([]byte(strings.TrimLeft(s[1], "0")), maskPos[:0])
			continue
		}
		addr, err := strconv.Atoi(s[0][4 : len(s[0])-1])
		if err != nil {
			log.Fatal(err)
		}
		num, err := strconv.Atoi(s[1])
		if err != nil {
			log.Fatal(err)
		}

		mem[addr] = (num & maskZeros) | maskOnes

		base := (addr & maskFloats) | maskOnes
		for _, i := range maskPos {
			mem2[base|i] = num
		}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	sum := 0
	for _, v := range mem {
		sum += v
	}
	fmt.Println("Part 1:", sum)

	sum2 := 0
	for _, v := range mem2 {
		sum2 += v
	}
	fmt.Println("Part 2:", sum2)
}

func processMask(b []byte, posMask []int) (int, int, int, []int) {
	pos := []int{}
	zeros := 0
	ones := 0
	fls := 0
	for n, i := range b {
		p := len(b) - n - 1
		k := 1 << p
		switch i {
		case '0':
			zeros |= k
		case '1':
			ones |= k
		case 'X':
			fls |= k
			pos = append(pos, k)
		}
	}
	return ^zeros, ones, ^fls, maskDfs(0, pos, posMask)
}

func maskDfs(base int, pos []int, posMask []int) []int {
	if len(pos) == 0 {
		return append(posMask, base)
	}
	posMask = maskDfs(base, pos[1:], posMask)
	return maskDfs(base|pos[0], pos[1:], posMask)
}

File: day14/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BErr = Box<dyn std::error::Error>;

fn main() -> Result<(), BErr> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut mem = HashMap::new();
    let mut mem2 = HashMap::new();

    let mut mask = (0, 0, 0);
    let mut mask_pos = Vec::new();

    for line in reader.lines() {
        let line = line?;
        let s = line.split(" = ").collect::<Vec<_>>();
        if s.len() != 2 {
            return Err("Invalid line".into());
        }
        if s[0] == "mask" {
            mask_pos.clear();
            let chars = s[1].chars().collect::<Vec<_>>();
            mask = process_mask(&chars, &mut mask_pos);
            continue;
        }
        let addr = s[0][4..s[0].len() - 1].parse::<usize>()?;
        let num = s[1].parse::<usize>()?;

        mem.insert(addr, (num & mask.0) | mask.1);

        let base = (addr & mask.2) | mask.1;
        for i in &mask_pos {
            mem2.insert(base | i, num);
        }
    }

    let sum = mem.values().sum::<usize>();
    println!("Part 1: {}", sum);

    let sum2 = mem2.values().sum::<usize>();
    println!("Part 2: {}", sum2);

    Ok(())
}

fn process_mask(b: &[char], pos_mask: &mut Vec<usize>) -> (usize, usize, usize) {
    let mut pos = Vec::new();
    let mut zeros = 0;
    let mut ones = 0;
    let mut fls = 0;
    let l = b.len();
    for (n, i) in b.iter().enumerate() {
        let p = l - n - 1;
        let k = 1 << p;
        match i {
            '0' => zeros |= k,
            '1' => ones |= k,
            'X' => {
                fls |= k;
                pos.push(k);
            }
            _ => (),
        }
    }
    mask_dfs(0, &pos, pos_mask);
    (!zeros, ones, !fls)
}

fn mask_dfs(base: usize, pos: &[usize], pos_mask: &mut Vec<usize>) {
    let (head, rest) = if let Some((head, rest)) = pos.split_first() {
        (head, rest)
    } else {
        pos_mask.push(base);
        return;
    };
    mask_dfs(base, rest, pos_mask);
    mask_dfs(base | head, rest, pos_mask);
}
